src/encoder.rs

```diff
@@ -1748,30 +1748,50 @@ pub fn motion_compensate(fi: &FrameInvariants, fs: &mut FrameState, cw: &mut Con
     let po = bo.plane_offset(&fs.input.planes[p].cfg);
     let rec = &mut fs.rec.planes[p];
     // TODO: make more generic to handle 2xN and Nx2 MC
-    if p > 0 && bsize == BlockSize::BLOCK_4X4 {
-      let some_use_intra = cw.bc.at(&bo.with_offset(-1,-1)).mode.is_intra()
-        || cw.bc.at(&bo.with_offset(0,-1)).mode.is_intra()
-        || cw.bc.at(&bo.with_offset(-1,0)).mode.is_intra();
+    if p > 0 && bsize < BlockSize::BLOCK_8X8 {
+      let mut some_use_intra = false;
+      if bsize == BlockSize::BLOCK_4X4 || bsize == BlockSize::BLOCK_4X8 {
+          some_use_intra |= cw.bc.at(&bo.with_offset(-1,0)).mode.is_intra(); };
+      if !some_use_intra && bsize == BlockSize::BLOCK_4X4 || bsize == BlockSize::BLOCK_8X4 {
+          some_use_intra |= cw.bc.at(&bo.with_offset(0,-1)).mode.is_intra(); };
+      if !some_use_intra && bsize == BlockSize::BLOCK_4X4 {
+          some_use_intra |= cw.bc.at(&bo.with_offset(-1,-1)).mode.is_intra(); };
 
       if some_use_intra {
         luma_mode.predict_inter(fi, p, &po, &mut rec.mut_slice(&po), plane_bsize.width(),
           plane_bsize.height(), ref_frames, mvs, bit_depth);
       } else {
         assert!(xdec == 1 && ydec == 1);
-        // TODO: these are only valid for 4:2:0
-        let mv0 = cw.bc.at(&bo.with_offset(-1,-1)).mv;
-        let rf0 = cw.bc.at(&bo.with_offset(-1,-1)).ref_frames;
-        let mv1 = cw.bc.at(&bo.with_offset(0,-1)).mv;
-        let rf1 = cw.bc.at(&bo.with_offset(0,-1)).ref_frames;
-        let po1 = PlaneOffset { x: po.x+2, y: po.y };
-        let mv2 = cw.bc.at(&bo.with_offset(-1,0)).mv;
-        let rf2 = cw.bc.at(&bo.with_offset(-1,0)).ref_frames;
-        let po2 = PlaneOffset { x: po.x, y: po.y+2 };
-        let po3 = PlaneOffset { x: po.x+2, y: po.y+2 };
-        luma_mode.predict_inter(fi, p, &po, &mut rec.mut_slice(&po), 2, 2, rf0, mv0, bit_depth);
-        luma_mode.predict_inter(fi, p, &po1, &mut rec.mut_slice(&po1), 2, 2, rf1, mv1, bit_depth);
-        luma_mode.predict_inter(fi, p, &po2, &mut rec.mut_slice(&po2), 2, 2, rf2, mv2, bit_depth);
-        luma_mode.predict_inter(fi, p, &po3, &mut rec.mut_slice(&po3), 2, 2, ref_frames, mvs, bit_depth);
+        // TODO: these are absolutely only valid for 4:2:0
+        if bsize == BlockSize::BLOCK_4X4 {
```

Comment:
There is quite a bit of code duplication in the following. Maybe this can be written as generic loops (over rows and columns) with a single call to predict_inter in the innermost loop?